src/scatter_buildblock/ScatterEstimation.cxx

```diff
@@ -0,0 +1,1486 @@
+
+    //Recompute the initial activity image if the max is equal to the min.
+#if SPEED_UP_FOR_DEBUG == 0
+    if( this->current_activity_image_sptr->find_max() == this->current_activity_image_sptr->find_min() )
+    {
+        info("ScatterEstimation: The max and the min values of the current activity image are equal."
+             "We deduct that it has been initialised to some value, therefore we will run the intial "
+             "recontruction ...");
+
+        if (iterative_method)
+            reconstruct_iterative(0, this->current_activity_image_sptr);
+        else
+            reconstruct_analytic(0, this->current_activity_image_sptr);
+
+        if ( run_debug_mode )
+        {
+            std::string out_filename = extras_path.get_path() + "inital_activity_image";
+            OutputFileFormat<DiscretisedDensity < 3, float > >::default_sptr()->
+                    write_to_file(out_filename, *this->current_activity_image_sptr);
+        }
+    }
+#else
+    {
+        std::string filename = extras_path.get_path() + "recon_0.hv";
+        current_activity_image_sptr = read_from_file<DiscretisedDensity<3,float> >(filename);
+    }
+#endif
+    // Set the first activity image
+    scatter_simulation_sptr->set_activity_image_sptr(current_activity_image_sptr);
+
+    if (this->do_average_at_2)
+        act_image_for_averaging.reset(this->current_activity_image_sptr->clone());
+
+    //
+    // Begin the estimation process...
+    //
+    info("ScatterEstimation: Begin the estimation process...");
+    for (int i_scat_iter = 1;
+         i_scat_iter <= this->num_scatter_iterations;
+         i_scat_iter++)
+    {
+
+
+        if ( this->do_average_at_2)
+        {
+            if (i_scat_iter == 2) // do average 0 and 1
+            {
+                if (is_null_ptr(act_image_for_averaging))
+                    error("Storing the first activity estimate has failed at some point.");
+
+                *this->current_activity_image_sptr += *act_image_for_averaging;
+                *this->current_activity_image_sptr /= 2.f;
+            }
+        }
+
+        if (!iterative_method)
+        {
+            // Threshold
+        }
+
+        info("ScatterEstimation: Scatter simulation in progress...");
+
+        if (this->scatter_simulation_sptr->process_data() == Succeeded::no)
+            error("ScatterEstimation: Scatter simulation failed");
+        info("ScatterEstimation: Scatter simulation done...");
+
+        if(this->run_debug_mode) // Write unscaled scatter sinogram
+        {
+            std::stringstream convert;   // stream used for the conversion
+            convert << "unscaled_" << i_scat_iter;
+            FilePath tmp(convert.str(),false);
+            tmp.prepend_directory_name(extras_path.get_path());
+            unscaled_est_projdata_sptr->write_to_file(tmp.get_string());
+        }
+
+        // Set the min and max scale factors
+        if (i_scat_iter > 0)
```

Comment:
```suggestion
        // We're going to assume that the first iteration starts from an image without scatter correction, and therefore
        // overestimates scatter. This could be inaccurate, but is the case most of the time.
        // TODO introduce a variable to control this behaviour
        if (i_scat_iter > 0)
```